### ISSUE TITLE
Getting error: "Could not connect to Twitter API using configured credentials." When adding new Twitter user.

### ISSUE THREAD

**Issue by mossiso** [gwu-libraries/social-feed-manager#299]
This is on a fresh, new installation of sfm. Following the instructions on readthedocs I get to the point to add the BBCNews user, but get the error mentioned.

I modified the ui/models.py file and the error is from this line (130-131):

```
if api is None:
            raise ValidationError('Could not connect to Twitter \
                                               API using configured credentials.')
```

I'm new to python and Django, so I'm not sure how to further diagnose the issue. It looks like the Twitter credentials are not being set in the authenticated_api method?

I'm using CentOS 6.6, mod_wsgi: Compiled for Python/2.7.8., mod_wsgi: Runtime using Python/2.7.9., Apache/2.2.15

**Comment by dchud** [comment]
Hi, this is one of the more confusing steps, and we haven't found a way to make it simpler yet, I'm sorry about that.

If this is happening, it's possible that the user configuration isn't all set up quite yet.  This is the tricky bit, there are three key kinds of users here:

 - the sfm-authorized app user
 - the twitter-authorized app user
 - TwitterUsers who tweets are collected by sfm

This is described somewhat clumsily (I wrote it) here:

  http://social-feed-manager.readthedocs.org/en/m5_002/install.html#first-time-running-sfm

The sfm-authorized app user ("sfmadmin" or whatever you called it) is what you use to add users, but isn't authenticated with Twitter unless you take the extra step to do that.  You will have to log in to sfm through Twitter using the same Twitter account you used to create the authorization keys.  This should be the exact same account name you set as the TWITTER_DEFAULT_USER value in the settings.

When you log in to sfm through twitter (you might have to log out as sfmadmin first), sfm saves authorization tokens for that user account in the system.  Then, when the API is called to add BBCNews or whatever other TwitterUser you want to collect, that TWITTER_DEFAULT_USER's tokens are used to authenticate the API calls.

When the error below occurs, it usually happens because whatever twitter-authorized sfm account is configured as sfm's TWITTER_DEFAULT_USER doesn't have those tokens stored in sfm yet.  It only takes the one time logging in to sfm through twitter with the TWITTER_DEFAULT_USER account to get those tokens; in our experience the tokens haven't expired yet.

If you've done all this and still get this error, there's something else that's wrong.  Could you please check that you have and confirm?  An easy way to do this is also in the docs:  look for the section about "User Social Auths" and check that the TWITTER_DEFAULT_USER has a social auth saved in sfm.

Another thing that might go wrong is if no valid URL is set as the Callback URL in Twitter's app authorization/configuration screen.  This is described near the bottom of the section here:

  http://social-feed-manager.readthedocs.org/en/m5_002/install.html#configuration

Let us know what you find.

**Comment by mossiso** [comment]
Thanks for getting back to me, and helping out with trouble shooting. I actually found your instructions clear and understandable (unless I've totally missed something in my steps below :) ). 

I've logged into sfm (http://domain.com/login) using our twitter account (the same that is used in the settings as TWITTER_DEFAULT_USERNAME). I click the "Login with Twitter" button, I get the redirect to http://api.twitter.com and successfully log in using the twitter account.

It looks like the oauth info is in the database. When I log out of sfm and then login on the admin side (http://domain.com/admin) with our sfmadmin account (different than the twitter account), I can see the twitter account in the "User social auths" section with this type of data in the "Extra data" field for the twitter account (I anonymized the actual tokens and id):

{"access_token": "oauth_token_secret=ZZZZZZZYYYYYYYXXXXXXXX&oauth_token=1111111-xxxxxxxxxxyyyyyyyyzzzzzzz", "id": 11111111}

These are the same values that Twitter shows as the Access Tokens for that Twitter account.

At this point I am still unable to add a twitter user in the admin side.

I also double checked that the Callback URL is set in Twitter's app. I actually could not create the app unless the Website and Calback URL fields were valid URLs.
